### PR TITLE
#5635: HIP: Add Overloads for parallel_scan with return value for ThreadVectorRange

### PR DESCRIPTION
**Pull request by thearusable**
Related to #5635 #6453

~Depends on #6235~ (merged)

### CODE REVIEW

**Review by masterleinad — CHANGES_REQUESTED** [review]
Please avoid copying the implementation from the overloads not returning the total.

**Comment by thearusable** [comment]
> Why does this pull request also touch `Cuda`?

I created the PRs as a chain which should be merged in particular order. This one is the last one currently created so it contains the changes from other ones. There will be at least two more PRs created.

**Comment by masterleinad** [comment]
Marking as draft since dependencies have not been merged.

**Review by masterleinad — APPROVED** [review]

File: core/src/HIP/Kokkos_HIP_Team.hpp [hunk 1]
```diff
@@ -797,7 +797,8 @@ parallel_scan(const Impl::ThreadVectorRangeBoundariesStruct<
     // exclusive scan -- the final accumulation
     // of i's val will be included in the second
     // closure call later.
-    if (i < loop_boundaries.end && threadIdx.x > 0) closure(i - 1, val, false);
+    if (i - 1 < loop_boundaries.end && threadIdx.x > 0)
+      closure(i - 1, val, false);
```
Comment (cz4rs) [inline core/src/HIP/Kokkos_HIP_Team.hpp:801]: See the discussion https://github.com/kokkos/kokkos/pull/6235#discussion_r1331460296 for a similar fix in CUDA backend.

**Review by fnrizzi — APPROVED** [review]

**Comment by fnrizzi** [comment]
gcc timed out , rest passes

File: core/unit_test/TestTeamVector.hpp [hunk 2]
```diff
@@ -1016,7 +1016,7 @@ struct checkScan {
 };
 }  // namespace VectorScanReducer
 
-#if !(defined(KOKKOS_IMPL_CUDA_CLANG_WORKAROUND) || defined(KOKKOS_ENABLE_HIP))
+#if !defined(KOKKOS_IMPL_CUDA_CLANG_WORKAROUND)
```
Comment (Rombur) [inline core/unit_test/TestTeamVector.hpp:1019]: If it works for HIP and CUDA nvcc, there is a high probability that it also works for CUDA clang. Not for this PR but I think it's worth to check if the still need this condition.

**Review by Rombur — APPROVED** [review]

**Review by crtrott — APPROVED** [review]

**Pull request merged**